handlers.go

```diff
@@ -495,3 +515,76 @@ func (a *autographer) handleGetAuthKeyIDs(w http.ResponseWriter, r *http.Request
 	w.WriteHeader(http.StatusOK)
 	w.Write(signerIDsJSON)
 }
+
+func (a *autographer) handleGetX5U(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		httpError(w, r, http.StatusMethodNotAllowed, "%s method not allowed; endpoint accepts GET only", r.Method)
+		return
+	}
+	if r.Body != nil {
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			httpError(w, r, http.StatusBadRequest, "failed to read request body: %s", err)
+			return
+		}
+		if len(body) > 0 {
+			httpError(w, r, http.StatusBadRequest, "endpoint received unexpected request body")
+			return
+		}
+	}
+
+	pathKeyID, ok := mux.Vars(r)["keyid"]
+	if !ok {
+		httpError(w, r, http.StatusInternalServerError, "route is improperly configured")
+		return
+	}
+	if !signer.IDFormatRegexp.MatchString(pathKeyID) {
+		httpError(w, r, http.StatusBadRequest, "keyid in URL path '%s' is invalid, it must match %s", pathKeyID, signer.IDFormat)
+		return
+	}
+
+	pathChainFile, ok := mux.Vars(r)["chainfile"]
+	if !ok {
+		httpError(w, r, http.StatusInternalServerError, "route is improperly configured")
+		return
+	}
+	if !signer.IDFormatRegexp.MatchString(pathKeyID) {
+		httpError(w, r, http.StatusBadRequest, "keyid in URL path '%s' is invalid, it must match %s", pathKeyID, signer.IDFormat)
+		return
+	}
+
+	// The handler regex should reject such paths, but let's be extra certain that we have been given
+	// a path without any attempt to escape the X5U upload directory.
+	if strings.Contains(pathChainFile, "/") || strings.Contains(pathChainFile, "\\") || strings.Contains(pathChainFile, "..") {
+		httpError(w, r, http.StatusBadRequest, "Invalid X5U file name '%s'", pathChainFile)
+		return
+	}
+
+	// Lookup the signer, and see if it has a local X5U chain upload location.
+	// Treat all errors as a 404 to avoid leaking unnecessary signer details as this
+	// endpoint has no authentication.
+	for _, s := range a.getSigners() {
+		config := s.Config()
+		if config.ID != pathKeyID {
+			continue
+		}
+
+		// Check if the signer has a chain upload location, and the requested file exists.
+		parsedURL, err := url.Parse(config.ChainUploadLocation)
+		if err != nil || parsedURL.Scheme != "file" {
+			break
+		}
+		chainFilePath := path.Join(parsedURL.Path, pathChainFile)
+		stat, err := os.Stat(chainFilePath)
+		if err != nil || !stat.Mode().IsRegular() {
+			break
+		}
+
+		// Serve files from the chain upload location
+		http.ServeFile(w, r, chainFilePath)
```

Comment:
I wrote all the rest of this review and then got here and thought:

Stepping back, could some of this work be done with [http.FileServer](https://pkg.go.dev/net/http#FileServer) instead? Like, in the routing layer, we iterate over all of the `getSigners` just once and create a `http.FileServer(http.Dir(chainUploadLocationDir))` per keyid with a `file://` chainUploadLocation? It would mean not having to do basically any of these checks since [http.Dir](https://pkg.go.dev/net/http#Dir) and FileServer do them for us. The small cost is it would serve the dir under it, but I think we're okay with that for dev and test.